### ISSUE TITLE
front_end should detect package_config.json by file content and not name

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/sdk#40864]
See the code here https://github.com/dart-lang/sdk/blob/5c869545f2b0572cfae350c34909a3ee52edacc9/pkg/front_end/lib/src/base/processed_options.dart#L587

This doesn't allow you to pass a package_config.json file by any name other than exactly `.dart_tool/package_config.json`. This will likely cause problems in blaze where we can't create overlapping files but have multiple dart_library rules in a package (all of which will need to create files like this.).

The format of the file should be determined by checking if it starts with a `{`, as is done in the [package_config](https://github.com/dart-lang/package_config/blob/fe0b48db500f0964a1f194c816b1f5f5e5e848f8/lib/src/package_config_io.dart#L41) package.

cc @davidmorgan @lrhn

**Comment by lrhn** [comment]
@jensjoha

**Comment by jensjoha** [comment]
That's not true.
If the filename is not `[...]/.dart_tool/package_config.json` it will see if there exists a file next to the file you specified called `.dart_tool/package_config.json` and use that if there is. If not it will use the file you specified, whether thats a json file or not.
So for this to be a problem you need a dictionary structure like this:
```
 - file_you_specified.ext
 - .dart_tool/package_config.json
```
where the `.dart_tool/package_config.json` is invalid/unrelated to the file you specified.

I don't want to have to read files we won't need to figure out if we'll need it or not.
In my mind, we either do this, or we always pick the file specified, even if it's a .packages file.

**Comment by lrhn** [comment]
That is not the specified behavior, though.

You should always read the specified file, and *only* if it is not a JSON file (assuming then that it is a `.packages` file) should you check whether there is a "better file". If it is a JSON file, you should use it.

This is an interim behavior until we stop generating and supporting `.packages` files.

**Comment by jensjoha** [comment]
[Your right](https://github.com/dart-lang/language/blob/master/accepted/future-releases/language-versioning/package-config-file-v2.md#specifying-the-file) --- I haven't noticed that before.

In my opinion that really should be changed.
a) It seems completely arbitrary.
b) That it starts with a `{` doesn't guarantee that its a json file
c) Even if it is a json file, it's not necessarily package_config json
(likely many more points can be made here)

It's already confusing that if I tell it to use one file it (can) actually uses another one instead (https://github.com/dart-lang/language/issues/772).
The current behavior where it does something based on the filename is somewhat easily explainable from the command line given though.
The specified behavior where you - looking at the command line - don't know what will happen is - in my opinion - strictly worse.

Please reconsider.

**Comment by davidmorgan** [comment]
The specified behaviour seems okay from my point of view. We will always pass `--packages` and will guarantee that that file exists, and detecting the format from the content will work fine.

The behaviour as currently implemented should mostly work as we won't create a `.dart_tool` folder intentionally. But, it might happen by accident. From that point of view, using the specified `--packages` file first would be better. Also, it would save us always having an existence check for a file we know won't (shouldn't) exist.

For google3 it's best if tools never go looking for files that were not explicitly specified, but I understand this is not the same as what other users want.

tl;dr: for google3, only ever using `--packages` is best; using `--packages` first is next best; using `--packages` after checking `.dart_tool` is least good. But all three should be workable.

Thanks.

**Comment by lrhn** [comment]
The specified behavior is intended to allow existing tools passing a `--packages` parameter pointing to a `.packages` file to keep working after other tools have moved to using `package_config.json` instead.

If  we don't do the redirection to a `package_config.json` file, a tool which invokes, say, the dart2js compiler with a `--packages something/.packages` would fail to work after language versioning is introduced. The compiler would potentially not have sufficient information to correctly compile the program. Not because the information is not there, but because the tool is not pointing at it.
We could choose to just make that compilation fail, likely with a useless warning.
Instead we try to see if there is an adjacent `package_config.json` file to use instead. It's fine to emit a warning at this point, but compilation should use the correct configuration file.

A tool which points directly to the `package_config.json` file should work correctly.


Starting with a `{` (or rather, the first *non-whitespace* character being a `{`) does not guarantee that it's a JSON file. It does guarantee that it's *not a `.packages` file*. A `.packages` file must start with a comment (`#`) or a package name (does not allow `{`). A `package_config.json` file must start with a `{` character. The goal is not to detect any other random kind of file, simply to distinguish between a directly specified file being a `.packages` file or a `package_config.json` file.

It's true that if you point to, say, a *Dart* file, it's going to be seen as a non-`package_config.json` file and we'll look for an adjacent `package_config.json` file even if the original was not even a `.packages` file. Pointing at random files gives odd behavior, but that's not the situation we're optimizing for.

**Comment by jensjoha** [comment]
Let me put it another way then: You're actively punishing people for using the default names (as generated by pub).

The three main ways to call it - I'm guessing - must be:
* specified with a .packages file with no .dart_tool/package_config.json file
* specified with a .packages file with a .dart_tool/package_config.json
* specified with a .dart_tool/package_config.json file.

Currently this will happen in the three situations:
* We'll check if .dart_tool/package_config.json exists, it doesn't, we'll check if the specified file exists, it does, we'll read that. 2 checks and 1 read.
* We'll check if .dart_tool/package_config.json exists, it does, we'll read that. 1 check and 1 read.
* We'll check if the specified file (already .dart_tool/package_config.json) exists, it does, we'll read that. 1 check and 1 read.

With the specified behavior:
* We'll check if the specified version exists. It does. We'll read it. It's not to our liking. We'll check if .dart_tool/package_config.json exists, it doesn't. Lets say we've optimized and still have the old read file in memory, so we'll reuse that. 2 checks and 1 read.
* We'll check if the specified version exists. It does. We'll read it. It's not to our liking. We'll check if .dart_tool/package_config.json exists, it does, we'll read that. 2 checks and 2 reads.
* We'll check if the specified file (already .dart_tool/package_config.json) exists, it does, we'll read it, it's to our liking. 1 check and 1 read.

As you can see the "middle case" (with a .packages file with a .dart_tool/package_config.json) - currently the default if you run pub I might add - becomes twice as costly.

This - apparently - to optimize for a non-standard situation where one specified a correct json file located in a wrong place and has a wrong json file located in the right place.

Am I really the only one who finds this odd?

**Comment by lrhn** [comment]
The behavior is "optimized" for the situation where you point to a `package_config.json` file.
That is the only long-term behavior that we need to support, and what we want tools which specify a file to migrate to as soon as possible.

If those tools point to a `.packages` file, then we will detect that (by reading the file) and then check if a better file as available in the default location relative to the default location of a `.package` file. 
This is the only situation where we assume anything about directory structures or file names. 
There is no requirement that any file has the default name or is placed in the default location.

One thing we could perhaps do instead is to *only* look for an alternative file if the originally provided file is named `.packages`. That would suggest that we are in a default layout, which is the only case where it makes sense to look for the other file in its default relative position.
Then we could even do that check before reading the `.packages` file.
How would that work?

**Comment by davidmorgan** [comment]
SGTM

**Comment by jensjoha** [comment]
Just making sure I'm reading you right:
* If the specified file is .packages, look in .dart_tool/package_config.json and use that if it exists (no matter what it might be)
* If the specified file is _not_ .packages, use that (no matter what it might be).

If I'm understanding you correctly, it sounds good to me to.

**Comment by lrhn** [comment]
OK, I propose that as a change then:

> When reading a package configuration file from a provided file path, 
> if that file path has a *file name* of `.packages` (with any path allowed before the file name), 
> then check if there exists a `./.dart_tool/package_config.json` file relative to the original file path.
> If so, read the `package_config.json` file instead. 
> It is an error if this other file is not a valid package configuration JSON file.
> If there is no such file, read the originally specified file.

WDYT @leafpetersen, @jonasfj?

**Comment by jensjoha** [comment]
`It is an error if this other file is not a valid package configuration JSON file.`
Why?

**Comment by lrhn** [comment]
Because picking that file is done under the assumption that we are hitting the default case where Pub has put a JSON file next to a `.packages` file. If that JSON file is actually a `.packages` file in disguise, the heuristic has failed. Arguably we are into malicious intent territory now, and we might not want to worry about it.

Feel free to recover from the error, as long as you tell the user they are doing something questionable.

**Comment by jensjoha** [comment]
I'd like a handle to tell package:package_config to give an error on non-new format then.

**Comment by jonasfj** [comment]
SGTM

**Comment by jakemac53** [comment]
I am fine with the proposed change to the semantics (treating `.packages` specially instead of `package_config.json` specially).

In general optimizing for a fixed number of between 1-3 file reads here is over-optimizing regardless though, and I think the current logic is also perfectly fine.

The _really_ important thing is that if you explicitly pass a package_config.json file (by any name, and nested under any directory), that it is used and parsed as a package_config.json file.  

> I'd like a handle to tell package:package_config to give an error on non-new format then.

It already does have `PackageConfig.parse{Bytes|String|Json}` methods which only work for a valid package_config.json file.

**Comment by jensjoha** [comment]
> I see no reason to worry about slightly slower compatibility mode.

> In general optimizing for a fixed number of between 1-3 file operations here is over-optimizing regardless though [...]

I definitely disagree.
Sure, this is only one place, and that in it self might not be measurable (at least on the big work machines we have) - but if the general idea is that "this doesn't matter" soon we'll have a compiler that's much slower than it needs to be.
Quite frankly the thought, in my opinion, should always be the other way around: Can I make this faster (even if it's just a little bit), and in this case it's obvious (at least to me) that unnecessarily reading files from a disk (a relatively slow operation) is going in the wrong direction.

> It already does have `PackageConfig.parse{Bytes|String|Json}` methods which only work for a valid package_config.json file.

You're right. I didn't see that.
